helix-core/src/test/java/org/apache/helix/integration/task/TestNoDoubleAssign.java

```diff
@@ -155,11 +155,16 @@ private void pollForDoubleAssign() {
    * Randomly causes Participants to lost connection temporarily.
    */
   private void breakConnection() {
+    // Note, send to THREAD_COUNT == 1 is a must to avoid leaking ClusterManager (participant).
```

Comment:
I suspect THREAD_COUNT > 1 is the test requirement here. But I'm not sure. @narendly, could you please comment here?

Reply:
This test relies on randomness. Also, the do not really wait for enough time to do breakConnection() etc. However, if we wait, there will be even a lot more leaking. Let us make an agreement that the design of the test is out of scope of this pull request. First stabilize the test. Will use singleThreadExecutor for now,

Reply:
@kaisun2000 Can you help me understand why THREAD_COUNT == 1 is a must to avoid leaking ClusterManager (participant)? Does it cause leakage even we shut down the executor after test?
In my understanding, the thread count there doesn't necessarily mean that this executor will use this number of threads at most. This number is for the core thread number, so more like a minimum number concept. If there's a  need, the executor will allocate more threads to handle the tasks even though THREAD_COUNT is specified as 1.

Reply:
Sure. Let me explain:

Assume we have two thread A, B. Both of them choose slot 0. The sequence of action is 

A stop 0
B stop 0
A start 0
B start 0

Now _participants[0] hold the reference of clutermanager started by B.  clustermanager started by A is not referenced by any one any more. They are leaked.

Reply:
Thanks for explaining @kaisun2000 . It makes sense.

Reply:
@kaisun2000 if that's the only reason, then we can just add a concurrent control on the stop, start block?

Moreover, I believe startParticipant will auto close it if one object is already running. We can also add synchronize there.

Reply:
something like this? I can give it a try. But it basically boils down to single thread case anyway. 

```
_executorServiceConnection.scheduleAtFixedRate(() -> {
      // Randomly pick a Participant and cause a transient connection issue
     synchronized (_executorServiceConnection) {
      int participantIndex = RANDOM.nextInt(_numNodes);
      stopParticipant(participantIndex);
      startParticipant(participantIndex);
      }
  }, 0L, CONNECTION_DELAY, TimeUnit.MILLISECONDS);

```

Reply:
> I believe startParticipant will auto close it if one object is already running. We can also add synchronize there.

Yes, it has. But over there still boils down to single thread execution. It does not increase parallism in this case. The above code seems to be clear in terms that we wan to stress that cleaning the resource by stopping it. Or we need to think about object life cycle.